### PR TITLE
fix: Improve handling of futures and threads during refresh.

### PR DESCRIPTION
**Pull request by hessjcg**
Rewrite the performRefresh() as a chain of task futures from the ListeningScheduledExecutorService. Now, tasks
submitted to the ListeningScheduledExecutorService never block on another task submitted to the ListeningScheduledExecutorService.

This should fix a category of bugs that show up in exceptions and logs as "connection timed out" or "refresh failed"
or "bad client certificate". These exceptions can occur when the credentials fail to refresh. 

This is the underlying bug: The ListeningScheduledExecutorService gets into a state where all its threads are busy 
running tasks, all running tasks are blocked waiting for recently submitted task to complete, and the recently
submitted tasks can't start because there are no available threads in the ListeningScheduledExecutorService.

### CODE REVIEW

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 1]
```diff
@@ -113,8 +116,19 @@ private InstanceData getInstanceData() {
       instanceDataFuture = currentInstanceData;
     }
     try {
-      return Uninterruptibles.getUninterruptibly(instanceDataFuture);
-    } catch (ExecutionException ex) {
+      return instanceDataFuture.get(30, TimeUnit.SECONDS);
```
Comment (hessjcg) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:119]: This is a small change of behavior: 

The old way hangs application thread until the current request finishes. If the request succeeds, this returns the InstanceData. If not, this throws a RuntimeException. 

The new way waits up to 30 seconds until a refresh attempt succeeds. If a refresh attempt succeeds within the 30 second timeout, the a future containing the latest InstanceData is returned. If not, this throws a RuntimeException using the exception from the last failed refresh attempt.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 2]
```diff
@@ -173,30 +187,63 @@ void forceRefresh() {
...
+   * futures that will 1. Acquire a rate limiter. 2. Attempt to fetch instance data. 3. Schedule the
+   * next attempt to get instance data based on the success/failure of this attempt.
+   *
+   * @see com.google.cloud.sql.core.CloudSqlInstance#rescheduleAfterGetInstanceData(
+   *     com.google.common.util.concurrent.ListenableFuture)
    */
-  private InstanceData performRefresh() throws InterruptedException, ExecutionException {
-    logger.fine(
-        String.format("[%s] Refresh Operation: Acquiring rate limiter permit.", instanceName));
+  private ListenableFuture<InstanceData> performRefresh() {
```
Comment (hessjcg) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:204]: This is a change of behavior. 

The old way: performRefresh() used to block until either (1) an InstanceData was retrieved or (2) an attempt failed, and an exception was thrown. 

The new way: performRefresh() does not block. It sets up a chain of futures. The future returned from this method will resolve only when a refresh attempt succeeds.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java [hunk 3]
```diff
@@ -111,12 +111,49 @@ public void testInstanceFailsOnConnectionError() throws Exception {
...
             keyPairFuture,
-            RateLimiter.create(1.0 / 30.0));
+            RateLimiter.create(10));
 
     RuntimeException ex = Assert.assertThrows(RuntimeException.class, instance::getSslData);
     assertThat(ex).hasMessageThat().contains("always fails");
   }
 
+  @Test
+  public void testInstanceFailsOnTooLongToRetrieve() throws Exception {
```
Comment (hessjcg) [inline core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java:121]: A new test covers the behavior when the refresh attempt never returns.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java [hunk 4]
```diff
@@ -139,7 +176,7 @@ public void testCloudSqlInstanceForcesRefresh() throws Exception {
             stubCredentialFactory,
             executorService,
             keyPairFuture,
-            RateLimiter.create(1.0 / 30.0));
+            RateLimiter.create(10));
```
Comment (hessjcg) [inline core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java:179]: Speed up tests - allow 10 requests/second.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java (hunk 1 above)
Comment (enocom) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:119]: Minor not: could we pull out the 30 into a constant? `MAX_INSTANCE_DATA_WAIT_TIME` or similar?

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 5]
```diff
@@ -216,13 +264,22 @@ private InstanceData performRefresh() throws InterruptedException, ExecutionExce
...
-            executor.schedule(this::performRefresh, secondsToRefresh, TimeUnit.SECONDS);
         // Refresh completed successfully, reset forceRefreshRunning.
         forceRefreshRunning = false;
+        currentRefreshFailure = null;
+        currentInstanceData = Futures.immediateFuture(data);
+
+        // Now update nextInstanceData to perform a refresh after the
+        // scheduled delay
+        ListenableFuture scheduleDelay =
+            executor.schedule(() -> {}, secondsToRefresh, TimeUnit.SECONDS);
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:275]: Is there any other ways to do this? On the surface it looks unusual.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java [hunk 6]
```diff
@@ -111,12 +111,49 @@ public void testInstanceFailsOnConnectionError() throws Exception {
...
+          ListenableFuture<?> f =
+              exec.submit(
+                  () -> {
+                    try {
+                      Thread.sleep(120 * 1000); // stop for 2 minutes
+                    } catch (InterruptedException e) {
+                    }
+                    throw new RuntimeException("fake read timeout");
+                  });
+          f.get(); // this will throw an ExecutionException
```
Comment (enocom) [inline core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java:138]: Let's add some more details to the comment -- it will throw an exception because we've configured the connector wait only 30s?

**Review by enocom — COMMENTED** [review]
This looks like a nice improvement.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java (hunk 1 above)
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java]: Done.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java (hunk 5 above)
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java]: I found a better method call for this.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java (hunk 6 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java]: I added comments and cleaned this up.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java [hunk 7]
```diff
@@ -119,18 +116,55 @@ public void testInstanceFailsOnConnectionError() throws Exception {
...
+    SslData sslData = new SslData(null, null, null);
+    InstanceData data =
+        new InstanceData(null, sslData, Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
     AtomicInteger refreshCount = new AtomicInteger();
 
     CloudSqlInstance instance =
         new CloudSqlInstance(
             "project:region:instance",
             (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+              Thread.sleep(100);
```
Comment (enocom) [inline core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java:165]: Can we avoid the sleep somehow? These are notoriously flaky otherwise.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 8]
```diff
@@ -99,23 +103,34 @@ class CloudSqlInstance {
...
+      synchronized (instanceDataGuard) {
+        if (currentRefreshFailure != null) {
+          throw new RuntimeException(
+              "Unable to get valid instance data within 30 seconds. Last refresh attempt failed:"
+                  + currentRefreshFailure.getMessage(),
+              currentRefreshFailure);
+        }
+      }
+      throw new RuntimeException(
+          "Unable to get valid instance data within 30 seconds. No refresh has completed.", e);
```
[the remark this answers is not in the document]
Comment (enocom) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:132]: Same above.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java (hunk 7 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java]: I've updated these tests (and added a few new ones) to use a lock instead of waiting.

**Comment by enocom** [comment]
@hessjcg Let's get the tests green and then we can get this merged

**Comment by enocom** [comment]
@hessjcg friendly ping on this. This is an important fix and improvement that we want to bake in CI to see if it resolves the flaky tests.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 9]
```diff
@@ -47,6 +47,8 @@ class CloudSqlInstance {
 
   private static final Logger logger = Logger.getLogger(CloudSqlInstance.class.getName());
 
+  public static final int MAX_INSTANCE_DATA_WAIT_MS = 30000;
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:50]: Let's add a comment explaining what this constant represents.

File: pom.xml [hunk 10]
```diff
@@ -153,6 +153,7 @@
           <compilerArgs>
             <arg>-XDcompilePolicy=simple</arg>
             <arg>-Xplugin:ErrorProne</arg>
+            <arg>-Xlint:-options</arg>
```
Comment (enocom) [inline pom.xml:156]: Should this be a separate PR?

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 11]
```diff
@@ -159,45 +216,88 @@ String getPreferredIp(List<String> preferredTypes) {
    * new refresh is already in progress. If successful, other methods will block until refresh has
    * been completed.
    */
-  void forceRefresh() {
+  boolean forceRefresh() {
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:219]: Should this be returning a boolean? I don't see it being used.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 12]
```diff
@@ -99,23 +104,71 @@ class CloudSqlInstance {
...
     synchronized (instanceDataGuard) {
       instanceDataFuture = currentInstanceData;
+      // If the currentInstanceData has expired, then force refresh (which will balk if a refresh
+      // is already running) and make this and future requests to getInstanceData wait on the
+      // refresh operation to complete.
+      if (instanceDataFuture.isDone()) {
+        Instant expiration;
+        try {
+          expiration = instanceDataFuture.get().getExpiration();
+          if (expiration == null || expiration.isBefore(Instant.now())) {
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:140]: If the current instance data has already expired, won't we ended up initiating a force refresh when the socket creation fails? In other words, do we really need to do this work here?

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 13]
```diff
@@ -99,23 +104,71 @@ class CloudSqlInstance {
...
+      }
     }
+
     try {
-      return Uninterruptibles.getUninterruptibly(instanceDataFuture);
-    } catch (ExecutionException ex) {
+      return instanceDataFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      synchronized (instanceDataGuard) {
+        if (currentRefreshFailure != null) {
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:156]: Could we split this out as a separate PR? It's nice to surface the refresh failure like this.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 14]
```diff
@@ -159,45 +216,88 @@ String getPreferredIp(List<String> preferredTypes) {
...
    */
-  void forceRefresh() {
+  boolean forceRefresh() {
     synchronized (instanceDataGuard) {
-      // Don't force a refresh until the current forceRefresh operation
+      // Don't force a refresh until the current refresh operation
       // has produced a successful refresh.
-      if (forceRefreshRunning) {
-        return;
+      if (refreshRunning) {
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:223]: Could this rename be a separate PR? Seems like a nice improvement.

**Comment by hessjcg** [comment]
Closing, too big. I'll submit this as a sequence of PRs.
